irma_to_miarka_file_lists.py

```diff
@@ -0,0 +1,797 @@
+    return end_date - datetime.timedelta(days=grace)
+
+
+def do_not_remove(fpath: str) -> bool:
+    # return true if a path is flagged with "_do_not_remove*"
+    fname = os.path.basename(fpath)
+    return "_do_not_remove" in fname.lower()
+
+
+def do_exclude(
+        fpath: str,
+        modification_date: datetime.datetime,
+        date_cutoff: datetime.datetime,
+        meta_path_and_dates: List[Tuple[str, datetime.datetime]]) -> bool:
+
+    msg = string.Template("${path} is ${action} because ${reason}")
+    reasons = [
+        f"{str(modification_date)} >= {str(date_cutoff)}",
+        f"{str(modification_date)} in meta data >= {str(date_cutoff)}",
+        "_do_not_remove in file name",
+        "_do_not_remove in companion file name",
+        "no condition to include file is met"
+    ]
+
+    # if disk modification date is after date_cutoff, return false
+    if modification_date >= date_cutoff:
+        print(msg.substitute(
+            path=fpath,
+            action="included",
+            reason=reasons[0]))
+        return False
+
+    fname = os.path.basename(fpath)
+    fdir = os.path.dirname(fpath)
+    meta_paths = [os.path.basename(p[0]) for p in meta_path_and_dates]
+    # if there is a modification date in the meta data after date cutoff, return false
+    try:
+        ix = meta_paths.index(fname)
+        if meta_path_and_dates[ix][1] >= date_cutoff:
+            print(msg.substitute(
+                path=fpath,
+                action="included",
+                reason=reasons[1]))
+            return False
+    except ValueError:
+        pass
+
+    # if the path is labeled with "_do_not_remove*", return false
+    if do_not_remove(fpath):
+        print(msg.substitute(
+            path=fpath,
+            action="included",
+            reason=reasons[2]))
+        return False
+
+    # if a path is accompanied with a "_do_not_remove*"-file on disk or in the meta data,
+    # return false
+    disk_paths = os.listdir(fdir) if os.path.exists(fdir) else []
+    do_not_remove_flags = list(filter(lambda p: do_not_remove(p), disk_paths + meta_paths))
+    # strip the "_do_not_remove*" suffix from the names
+    do_not_remove_names = [p[0:p.lower().index("_do_not_remove")] for p in do_not_remove_flags]
+    if fname in do_not_remove_names:
+        print(msg.substitute(
+            path=fpath,
+            action="included",
+            reason=reasons[3]))
+        return False
+
+    # the path should not be kept because of date or "_do_not_remove*"-indicator, return True
+    print(msg.substitute(
+        path=fpath,
+        action="excluded",
+        reason=reasons[4]))
+    return True
+
+
+def include_exclude(
+        search_path: str,
+        path_and_dates: List[Tuple[str, datetime.datetime]],
+        date_cutoff: datetime.datetime,
+        meta_path_and_dates: List[Tuple[str, datetime.datetime]],
+        exclude_fn: Callable = do_exclude) -> List[List[str]]:
```

Comment:
yes, my idea was that since that function really is the crucial point where the decisions are made and everything else is more or less scaffolding, it could be nice to be able to slot in alternative functionality. It's not used right now though 😄